internal/service/worker/controller.go

```diff
@@ -31,39 +31,173 @@ func GrpcRegister() {
-	var workers []*workerV1beta1.Worker
-	for k, v := range c.states {
-		if v == request.GetState() {
-			workers = append(workers, &workerV1beta1.Worker{
-				WorkerId: k,
-				State:    v,
-			})
+	pageSize := request.GetPageSize()
+	pageToken := request.GetPageToken()
+	st := request.GetStatus()
+	var v validate.Validator
+	v.ValidateString("page_token", pageToken, validate.IsNotNull)
+	v.Validate("page_size", pageSize, ValidatePageSize)
+	v.Validate("status", st, ValidateStatus)
+
+	var s Status
+	if st != workerV1beta1.WorkerStatus_WORKER_STATUS_UNSPECIFIED {
+		s = transformStatusFrom(st)
+	}
+	//todo make use of page_token as offset
+	all, err := QueryAll(ctx, &s, int(pageSize), 0)
+
+	if err != nil {
+		st := status.Newf(codes.Internal, "query all error: %s", err)
+		st, err = st.WithDetails(request)
+		if err != nil {
+			panic(fmt.Errorf("enrich grpc status with details error: %w", err))
 		}
+		return nil, st.Err()
+	}
+
+	var workers = make([]*workerV1beta1.Worker, 0, int(pageSize))
+	for _, v := range all {
+		workers = append(workers, transformWorker(v))
 	}
 
 	return &workerV1beta1.ListWorkersResponse{
-		Workers: workers,
+		Workers:       workers,
+		NextPageToken: "todo",
 	}, nil
 }
+
+var statusTo = map[Status]workerV1beta1.WorkerStatus{
+	Offline:  workerV1beta1.WorkerStatus_WORKER_STATUS_OFFLINE,
+	Ready:    workerV1beta1.WorkerStatus_WORKER_STATUS_READY,
+	OnJob:    workerV1beta1.WorkerStatus_WORKER_STATUS_ON_JOB,
+	Paused:   workerV1beta1.WorkerStatus_WORKER_STATUS_PAUSED,
+	Disabled: workerV1beta1.WorkerStatus_WORKER_STATUS_DISABLED,
+}
+
+var statusFrom = map[workerV1beta1.WorkerStatus]Status{
+	workerV1beta1.WorkerStatus_WORKER_STATUS_OFFLINE:  Offline,
+	workerV1beta1.WorkerStatus_WORKER_STATUS_READY:    Ready,
+	workerV1beta1.WorkerStatus_WORKER_STATUS_ON_JOB:   OnJob,
+	workerV1beta1.WorkerStatus_WORKER_STATUS_PAUSED:   Paused,
+	workerV1beta1.WorkerStatus_WORKER_STATUS_DISABLED: Disabled,
+}
+
+func transformWorker(wrk *Worker) *workerV1beta1.Worker {
+	s, ok := statusTo[wrk.Status]
+	if !ok {
+		s = workerV1beta1.WorkerStatus_WORKER_STATUS_UNSPECIFIED
+	}
+	return &workerV1beta1.Worker{
+		WorkerId: wrk.ID,
+		Status:   s,
+	}
+}
+
+func transformStatusFrom(sta workerV1beta1.WorkerStatus) Status {
+	s, ok := statusFrom[sta]
+	if !ok {
+		s = Offline
+	}
+	return s
+}
+
+func ValidateStatus(value interface{}) error {
+	v, ok := value.(workerV1beta1.WorkerStatus)
+	if !ok {
+		return fmt.Errorf("not a status value: %v", value)
+	}
+	switch v {
+	case workerV1beta1.WorkerStatus_WORKER_STATUS_OFFLINE:
+	case workerV1beta1.WorkerStatus_WORKER_STATUS_READY:
+	case workerV1beta1.WorkerStatus_WORKER_STATUS_ON_JOB:
+	case workerV1beta1.WorkerStatus_WORKER_STATUS_PAUSED:
+	case workerV1beta1.WorkerStatus_WORKER_STATUS_DISABLED:
+		break
+	default:
+		return fmt.Errorf("unknown status value: %v", value)
+	}
+	return nil
+}
```

Comment:
What's the purpose of enumerating them all out like that versus just using [`ozzo.Required`](https://pkg.go.dev/github.com/go-ozzo/ozzo-validation#Required)?

Reply:
If new additions are made to the enum, it's considered a non-breaking / backwards-compatible change.

That should be okay, as long as we always update the server first. (Which is pretty much a guarantee since the protos live in the same repo as server code).

Reply:
The 0th ordinal (WORKER_STATUS_UNSPECIFIED) is always the default if the client leaves the enum value blank or if it sends a newer, unknown/unrecognized value.

Since an enum is basically just an int32, [`ozzo.Required`](https://pkg.go.dev/github.com/go-ozzo/ozzo-validation#Required) will throw an error in either of those cases.

Reply:
Ok, i will simplify it to Required.

Reply:
@kevinmichaelchen
ValidateStatus should not allow UNSPECIFIED status to pass through, as Controller API is explicit about status payload. 
I left it in place, with doc comments.

Reply:
I was unaware we weren't using [this library](https://pkg.go.dev/github.com/go-ozzo/ozzo-validation).

Ideally we wouldn't have to update this function every time we add an enum value.

Reply:
Overall, no need to enumerate the entire enum in Go code like this. The validation function for an enum should basically boil down to: "This enum (int32) should not be equal to zero. Any value higher than that is allowed".

Reply:
thank you @kevinmichaelchen , fixed.